File: ios/Sources/Library/Services/QueueManager.swift
import Foundation
import Combine

// MARK: - Queue Manager

/// Manages the listening queue with persistence and automatic advancement.
@MainActor
final class QueueManager: ObservableObject {

    // MARK: - Published State

    @Published private(set) var state: QueueState = .empty
    @Published private(set) var isLoading: Bool = false

    // MARK: - Properties

    private let persistenceKey = "ReadAloudAI.QueueState"
    private let historyKey = "ReadAloudAI.ListeningHistory"
    private let maxHistoryItems = 100

    private var history: [ListeningHistoryEntry] = []
    private var currentHistoryEntry: ListeningHistoryEntry?
    private var autoAdvanceEnabled: Bool = true

    // Callbacks
    var onItemCompleted: ((QueueItem) -> Void)?
    var onQueueExhausted: (() -> Void)?
    var onCurrentItemChanged: ((QueueItem?) -> Void)?

    // MARK: - Singleton

    static let shared = QueueManager()

    // MARK: - Computed Properties

    var currentItem: QueueItem? {
        state.currentItem
    }

    var items: [QueueItem] {
        state.items
    }

    var upNext: [QueueItem] {
        state.upNext
    }

    var isEmpty: Bool {
        state.isEmpty
    }

    var count: Int {
        state.items.count
    }

    var hasNext: Bool {
        state.hasNext
    }

    var hasPrevious: Bool {
        state.hasPrevious
    }

    var repeatMode: RepeatMode {
        get { state.repeatMode }
        set {
            state.repeatMode = newValue
            saveState()
        }
    }

    var shuffleMode: ShuffleMode {
        get { state.shuffleMode }
        set {
            state.shuffleMode = newValue
            if newValue == .on {
                generateShuffledIndices()
            } else {
                state.shuffledIndices = nil
            }
            saveState()
        }
    }

    // MARK: - Initialization

    private init() {
        loadState()
        loadHistory()
    }

    // MARK: - Queue Operations

    /// Add an item to the queue.
    func addToQueue(_ item: QueueItem, position: QueuePosition = .last) {
        var newItem = item

        switch position {
        case .next:
            if let currentIndex = state.currentIndex {
                let insertIndex = currentIndex + 1
                state.items.insert(newItem, at: min(insertIndex, state.items.count))
            } else {
                state.items.insert(newItem, at: 0)
            }

        case .last:
            state.items.append(newItem)

        case .first:
            state.items.insert(newItem, at: 0)
            // Adjust current index if needed
            if let currentIndex = state.currentIndex {
                state.currentIndex = currentIndex + 1
            }

        case .at(let index):
            let safeIndex = max(0, min(index, state.items.count))
            state.items.insert(newItem, at: safeIndex)
            // Adjust current index if needed
            if let currentIndex = state.currentIndex, safeIndex <= currentIndex {
                state.currentIndex = currentIndex + 1
            }
        }

        if shuffleMode == .on {
            generateShuffledIndices()
        }

        saveState()
    }

    /// Add multiple items to the queue.
    func addToQueue(_ items: [QueueItem], position: QueuePosition = .last) {
        for (index, item) in items.enumerated() {
            switch position {
            case .next:
                addToQueue(item, position: .at((state.currentIndex ?? -1) + 1 + index))
            case .last:
                addToQueue(item, position: .last)
            case .first:
                addToQueue(item, position: .at(index))
            case .at(let baseIndex):
                addToQueue(item, position: .at(baseIndex + index))
            }
        }
    }

    /// Add an article to the queue.
    func addToQueue(_ article: Article, position: QueuePosition = .last) {
        let item = QueueItem(from: article)
        addToQueue(item, position: position)
    }

    /// Remove an item from the queue.
    func removeFromQueue(at index: Int) {
        guard index < state.items.count else { return }

        state.items.remove(at: index)

        // Adjust current index
        if let currentIndex = state.currentIndex {
            if index < currentIndex {
                state.currentIndex = currentIndex - 1
            } else if index == currentIndex {
                // Current item was removed
                if state.items.isEmpty {
                    state.currentIndex = nil
                } else if currentIndex >= state.items.count {
                    state.currentIndex = state.items.count - 1
                }
                onCurrentItemChanged?(currentItem)
            }
        }

        if shuffleMode == .on {
            generateShuffledIndices()
        }

        saveState()
    }

    /// Remove an item by ID.
    func removeFromQueue(id: UUID) {
        if let index = state.items.firstIndex(where: { $0.id == id }) {
            removeFromQueue(at: index)
        }
    }

    /// Remove multiple items.
    func removeFromQueue(at offsets: IndexSet) {
        // Remove in reverse order to maintain indices
        for index in offsets.sorted().reversed() {
            removeFromQueue(at: index)
        }
    }

    /// Move items in the queue.
    func moveItems(from source: IndexSet, to destination: Int) {
        let currentItemID = currentItem?.id

        state.items.move(fromOffsets: source, toOffset: destination)

        // Update current index to follow the current item
        if let id = currentItemID {
            state.currentIndex = state.items.firstIndex(where: { $0.id == id })
        }

        if shuffleMode == .on {
            generateShuffledIndices()
        }

        saveState()
    }

    /// Clear the entire queue.
    func clearQueue() {
        // Complete current history entry
        completeHistoryEntry()

        state = .empty
        saveState()
        onQueueExhausted?()
    }

    /// Clear completed items from the queue.
    func clearCompletedItems() {
        let currentID = currentItem?.id
        state.items.removeAll { $0.isCompleted }

        if let id = currentID {
            state.currentIndex = state.items.firstIndex(where: { $0.id == id })
        } else if !state.items.isEmpty {
            state.currentIndex = 0
        } else {
            state.currentIndex = nil
        }

        if shuffleMode == .on {
            generateShuffledIndices()
        }

        saveState()
    }

    // MARK: - Playback Control

    /// Start playing a specific item in the queue.
    func play(at index: Int) {
        guard index < state.items.count else { return }

        // Complete previous history entry
        completeHistoryEntry()

        state.currentIndex = index
        startHistoryEntry(for: state.items[index])

        saveState()
        onCurrentItemChanged?(currentItem)
    }

    /// Start playing a specific item by ID.
    func play(id: UUID) {
        if let index = state.items.firstIndex(where: { $0.id == id }) {
            play(at: index)
        }
    }

    /// Move to the next item in the queue.
    /// Returns the next item, or nil if queue is exhausted.
    @discardableResult
    func next() -> QueueItem? {
        guard let currentIndex = state.currentIndex else {
            if !state.items.isEmpty {
                play(at: 0)
                return currentItem
            }
            return nil
        }

        // Mark current as completed
        markCurrentAsCompleted()

        let nextIndex: Int?

        if shuffleMode == .on, let shuffled = state.shuffledIndices {
            let shufflePosition = shuffled.firstIndex(of: currentIndex) ?? currentIndex
            if shufflePosition + 1 < shuffled.count {
                nextIndex = shuffled[shufflePosition + 1]
            } else if repeatMode == .all {
                generateShuffledIndices()
                nextIndex = state.shuffledIndices?.first
            } else {
                nextIndex = nil
            }
        } else {
            if currentIndex + 1 < state.items.count {
                nextIndex = currentIndex + 1
            } else if repeatMode == .all {
                nextIndex = 0
            } else {
                nextIndex = nil
            }
        }

        if let next = nextIndex {
            play(at: next)
            return currentItem
        } else {
            state.currentIndex = nil
            saveState()
            onQueueExhausted?()
            return nil
        }
    }

    /// Move to the previous item in the queue.
    /// Returns the previous item, or nil if at the beginning.
    @discardableResult
    func previous() -> QueueItem? {
        guard let currentIndex = state.currentIndex else { return nil }

        let prevIndex: Int?

        if shuffleMode == .on, let shuffled = state.shuffledIndices {
            let shufflePosition = shuffled.firstIndex(of: currentIndex) ?? currentIndex
            if shufflePosition > 0 {
                prevIndex = shuffled[shufflePosition - 1]
            } else if repeatMode == .all {
                prevIndex = shuffled.last
            } else {
                prevIndex = nil
            }
        } else {
            if currentIndex > 0 {
                prevIndex = currentIndex - 1
            } else if repeatMode == .all {
                prevIndex = state.items.count - 1
            } else {
                prevIndex = nil
            }
        }

        if let prev = prevIndex {
            play(at: prev)
            return currentItem
        }

        return nil
    }

    /// Handle item completion - called when playback finishes.
    func itemCompleted() {
        guard let item = currentItem else { return }

        markCurrentAsCompleted()
        onItemCompleted?(item)

        if repeatMode == .one {
            // Replay current item
            if let index = state.currentIndex {
                state.items[index].lastPlayedPosition = 0
                saveState()
            }
        } else if autoAdvanceEnabled {
            next()
        }
    }

    /// Update playback position for current item.
    func updatePosition(_ position: TimeInterval) {
        guard let index = state.currentIndex else { return }

        state.items[index].lastPlayedPosition = position

        // Update history entry
        if var entry = currentHistoryEntry {
            entry.durationListened = position
            currentHistoryEntry = entry
        }

        // Save periodically (every 10 seconds)
        let positionSeconds = Int(position)
        if positionSeconds % 10 == 0 {
            saveState()
        }
    }

    // MARK: - Repeat & Shuffle

    /// Toggle repeat mode.
    func toggleRepeatMode() {
        repeatMode = repeatMode.next()
    }

    /// Toggle shuffle mode.
    func toggleShuffleMode() {
        shuffleMode.toggle()
    }

    // MARK: - Queue from Playlist

    /// Replace queue with playlist contents.
    func playPlaylist(_ playlist: Playlist, startingAt index: Int = 0) {
        clearQueue()

        let queueItems = playlist.items.enumerated().map { idx, item in
            QueueItem(
                articleID: item.articleID,
                title: item.title,
                author: item.author,
                siteName: item.siteName,
                sourceType: item.sourceType,
                duration: item.duration,
                heroImageURL: item.heroImageURL
            )
        }

        state.items = queueItems

        if shuffleMode == .on {
            generateShuffledIndices()
        }

        if !queueItems.isEmpty {
            play(at: index)
        }
    }

    /// Add playlist contents to queue.
    func addPlaylistToQueue(_ playlist: Playlist, position: QueuePosition = .last) {
        let queueItems = playlist.items.map { item in
            QueueItem(
                articleID: item.articleID,
                title: item.title,
                author: item.author,
                siteName: item.siteName,
                sourceType: item.sourceType,
                duration: item.duration,
                heroImageURL: item.heroImageURL
            )
        }

        addToQueue(queueItems, position: position)
    }

    // MARK: - History

    /// Get listening history.
    func getHistory(limit: Int = 50) -> [ListeningHistoryEntry] {
        Array(history.prefix(limit))
    }

    /// Clear listening history.
    func clearHistory() {
        history.removeAll()
        saveHistory()
    }

    // MARK: - Private Methods

    private func markCurrentAsCompleted() {
        guard let index = state.currentIndex else { return }

        state.items[index].isCompleted = true
        state.items[index].completedAt = Date()

        completeHistoryEntry()
        saveState()
    }

    private func generateShuffledIndices() {
        var indices = Array(0..<state.items.count)

        // Keep current item at current position if playing
        if let currentIndex = state.currentIndex {
            indices.remove(at: currentIndex)
            indices.shuffle()
            indices.insert(currentIndex, at: currentIndex)
        } else {
            indices.shuffle()
        }

        state.shuffledIndices = indices
    }

    private func startHistoryEntry(for item: QueueItem) {
        currentHistoryEntry = ListeningHistoryEntry(from: item)
    }

    private func completeHistoryEntry() {
        guard var entry = currentHistoryEntry else { return }

        entry.endedAt = Date()
        entry.completed = (currentItem?.isCompleted ?? false)

        history.insert(entry, at: 0)

        // Trim history
        if history.count > maxHistoryItems {
            history = Array(history.prefix(maxHistoryItems))
        }

        saveHistory()
        currentHistoryEntry = nil
    }

    // MARK: - Persistence

    private func saveState() {
        do {
            let data = try JSONEncoder().encode(state)
            UserDefaults.standard.set(data, forKey: persistenceKey)
        } catch {
            print("Failed to save queue state: \(error)")
        }
    }

    private func loadState() {
        guard let data = UserDefaults.standard.data(forKey: persistenceKey) else { return }

        do {
            state = try JSONDecoder().decode(QueueState.self, from: data)
        } catch {
            print("Failed to load queue state: \(error)")
            state = .empty
        }
    }

    private func saveHistory() {
        do {
            let data = try JSONEncoder().encode(history)
            UserDefaults.standard.set(data, forKey: historyKey)
        } catch {
            print("Failed to save history: \(error)")
        }
    }

    private func loadHistory() {
        guard let data = UserDefaults.standard.data(forKey: historyKey) else { return }

        do {
            history = try JSONDecoder().decode([ListeningHistoryEntry].self, from: data)
        } catch {
            print("Failed to load history: \(error)")
            history = []
        }
    }
}

// MARK: - Queue Manager Extensions

extension QueueManager {

    /// Get queue summary for display.
    var summary: String {
        let count = state.items.count
        if count == 0 {
            return "Queue is empty"
        }

        let totalMinutes = Int(state.totalDuration) / 60
        if totalMinutes < 60 {
            return "\(count) item\(count == 1 ? "" : "s") • \(totalMinutes) min"
        } else {
            let hours = totalMinutes / 60
            let mins = totalMinutes % 60
            return "\(count) item\(count == 1 ? "" : "s") • \(hours) hr \(mins) min"
        }
    }

    /// Get remaining time summary.
    var remainingSummary: String {
        let remaining = state.remainingDuration
        let minutes = Int(remaining) / 60

        if minutes < 60 {
            return "\(minutes) min remaining"
        } else {
            let hours = minutes / 60
            let mins = minutes % 60
            return "\(hours) hr \(mins) min remaining"
        }
    }

    /// Check if an article is in the queue.
    func contains(articleID: UUID) -> Bool {
        state.items.contains { $0.articleID == articleID }
    }

    /// Get position in queue for an article.
    func position(for articleID: UUID) -> Int? {
        state.items.firstIndex { $0.articleID == articleID }
    }
}

File: android/app/src/main/java/com/listenai/service/playback/QueueManager.kt
package com.listenai.service.playback

import com.listenai.data.models.Article
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.UUID

/**
 * Queue item representing an article in the listening queue
 */
data class QueueItem(
    val id: UUID = UUID.randomUUID(),
    val article: Article,
    val addedAt: Long = System.currentTimeMillis(),
    val position: Int = 0
)

/**
 * Manages the listening queue for articles.
 * Handles queue ordering, persistence, and playback coordination.
 */
class QueueManager {

    // Queue state
    private val _queue = MutableStateFlow<List<QueueItem>>(emptyList())
    val queue: StateFlow<List<QueueItem>> = _queue.asStateFlow()

    // Current playing item
    private val _currentItem = MutableStateFlow<QueueItem?>(null)
    val currentItem: StateFlow<QueueItem?> = _currentItem.asStateFlow()

    // Current index in queue
    private val _currentIndex = MutableStateFlow(-1)
    val currentIndex: StateFlow<Int> = _currentIndex.asStateFlow()

    // Shuffle and repeat modes
    private val _isShuffleEnabled = MutableStateFlow(false)
    val isShuffleEnabled: StateFlow<Boolean> = _isShuffleEnabled.asStateFlow()

    private val _repeatMode = MutableStateFlow(RepeatMode.OFF)
    val repeatMode: StateFlow<RepeatMode> = _repeatMode.asStateFlow()

    // Original order for shuffle restoration
    private var originalOrder: List<QueueItem> = emptyList()

    enum class RepeatMode {
        OFF,        // No repeat
        ALL,        // Repeat entire queue
        ONE         // Repeat current item
    }

    // MARK: - Queue Management

    /**
     * Add an article to the end of the queue
     */
    fun addToQueue(article: Article) {
        val newItem = QueueItem(
            article = article,
            position = _queue.value.size
        )
        _queue.value = _queue.value + newItem

        // If nothing is playing, set this as current
        if (_currentItem.value == null) {
            _currentItem.value = newItem
            _currentIndex.value = _queue.value.size - 1
        }
    }

    /**
     * Add an article to play next (after current item)
     */
    fun playNext(article: Article) {
        val newItem = QueueItem(
            article = article,
            position = _currentIndex.value + 1
        )

        val currentList = _queue.value.toMutableList()
        val insertIndex = (_currentIndex.value + 1).coerceIn(0, currentList.size)
        currentList.add(insertIndex, newItem)

        // Update positions
        _queue.value = currentList.mapIndexed { index, item ->
            item.copy(position = index)
        }
    }

    /**
     * Remove an item from the queue
     */
    fun removeFromQueue(itemId: UUID) {
        val currentList = _queue.value.toMutableList()
        val removeIndex = currentList.indexOfFirst { it.id == itemId }

        if (removeIndex >= 0) {
            currentList.removeAt(removeIndex)

            // Update current index if needed
            if (removeIndex < _currentIndex.value) {
                _currentIndex.value -= 1
            } else if (removeIndex == _currentIndex.value) {
                // Current item was removed, move to next or previous
                if (currentList.isNotEmpty()) {
                    val newIndex = removeIndex.coerceAtMost(currentList.size - 1)
                    _currentIndex.value = newIndex
                    _currentItem.value = currentList.getOrNull(newIndex)
                } else {
                    _currentIndex.value = -1
                    _currentItem.value = null
                }
            }

            // Update positions
            _queue.value = currentList.mapIndexed { index, item ->
                item.copy(position = index)
            }
        }
    }

    /**
     * Clear the entire queue
     */
    fun clearQueue() {
        _queue.value = emptyList()
        _currentItem.value = null
        _currentIndex.value = -1
        originalOrder = emptyList()
    }

    /**
     * Move an item to a new position in the queue
     */
    fun moveItem(fromIndex: Int, toIndex: Int) {
        val currentList = _queue.value.toMutableList()

        if (fromIndex in currentList.indices && toIndex in currentList.indices) {
            val item = currentList.removeAt(fromIndex)
            currentList.add(toIndex, item)

            // Update current index if affected
            when {
                fromIndex == _currentIndex.value -> _currentIndex.value = toIndex
                fromIndex < _currentIndex.value && toIndex >= _currentIndex.value -> _currentIndex.value -= 1
                fromIndex > _currentIndex.value && toIndex <= _currentIndex.value -> _currentIndex.value += 1
            }

            // Update positions
            _queue.value = currentList.mapIndexed { index, item ->
                item.copy(position = index)
            }
        }
    }

    // MARK: - Playback Control

    /**
     * Set the current playing item by index
     */
    fun playAtIndex(index: Int) {
        if (index in _queue.value.indices) {
            _currentIndex.value = index
            _currentItem.value = _queue.value[index]
        }
    }

    /**
     * Set the current playing item by ID
     */
    fun playItem(itemId: UUID) {
        val index = _queue.value.indexOfFirst { it.id == itemId }
        if (index >= 0) {
            playAtIndex(index)
        }
    }

    /**
     * Move to next item in queue
     * Returns true if there is a next item, false if at end
     */
    fun next(): Boolean {
        return when (_repeatMode.value) {
            RepeatMode.ONE -> {
                // Stay on current item
                true
            }
            RepeatMode.ALL -> {
                val nextIndex = (_currentIndex.value + 1) % _queue.value.size
                playAtIndex(nextIndex)
                true
            }
            RepeatMode.OFF -> {
                val nextIndex = _currentIndex.value + 1
                if (nextIndex < _queue.value.size) {
                    playAtIndex(nextIndex)
                    true
                } else {
                    false
                }
            }
        }
    }

    /**
     * Move to previous item in queue
     * Returns true if there is a previous item
     */
    fun previous(): Boolean {
        return when (_repeatMode.value) {
            RepeatMode.ONE -> {
                // Stay on current item
                true
            }
            RepeatMode.ALL -> {
                val prevIndex = if (_currentIndex.value <= 0) {
                    _queue.value.size - 1
                } else {
                    _currentIndex.value - 1
                }
                playAtIndex(prevIndex)
                true
            }
            RepeatMode.OFF -> {
                val prevIndex = _currentIndex.value - 1
                if (prevIndex >= 0) {
                    playAtIndex(prevIndex)
                    true
                } else {
                    false
                }
            }
        }
    }

    // MARK: - Shuffle and Repeat

    /**
     * Toggle shuffle mode
     */
    fun toggleShuffle() {
        _isShuffleEnabled.value = !_isShuffleEnabled.value

        if (_isShuffleEnabled.value) {
            // Save original order and shuffle
            originalOrder = _queue.value.toList()
            val currentItem = _currentItem.value
            val otherItems = _queue.value.filter { it.id != currentItem?.id }.shuffled()

            // Keep current item at front, shuffle rest
            _queue.value = if (currentItem != null) {
                listOf(currentItem) + otherItems
            } else {
                otherItems
            }.mapIndexed { index, item -> item.copy(position = index) }

            _currentIndex.value = 0
        } else {
            // Restore original order
            val currentItemId = _currentItem.value?.id
            _queue.value = originalOrder.mapIndexed { index, item -> item.copy(position = index) }
            _currentIndex.value = _queue.value.indexOfFirst { it.id == currentItemId }.coerceAtLeast(0)
        }
    }

    /**
     * Cycle through repeat modes: OFF -> ALL -> ONE -> OFF
     */
    fun cycleRepeatMode() {
        _repeatMode.value = when (_repeatMode.value) {
            RepeatMode.OFF -> RepeatMode.ALL
            RepeatMode.ALL -> RepeatMode.ONE
            RepeatMode.ONE -> RepeatMode.OFF
        }
    }

    /**
     * Set repeat mode directly
     */
    fun setRepeatMode(mode: RepeatMode) {
        _repeatMode.value = mode
    }

    // MARK: - Queue Info

    /**
     * Check if there's a next item available
     */
    fun hasNext(): Boolean {
        return when (_repeatMode.value) {
            RepeatMode.ONE, RepeatMode.ALL -> _queue.value.isNotEmpty()
            RepeatMode.OFF -> _currentIndex.value < _queue.value.size - 1
        }
    }

    /**
     * Check if there's a previous item available
     */
    fun hasPrevious(): Boolean {
        return when (_repeatMode.value) {
            RepeatMode.ONE, RepeatMode.ALL -> _queue.value.isNotEmpty()
            RepeatMode.OFF -> _currentIndex.value > 0
        }
    }

    /**
     * Get total queue duration in milliseconds
     */
    fun totalDuration(): Long {
        return _queue.value.sumOf { it.article.estimatedDuration }
    }

    /**
     * Get remaining queue duration from current position in milliseconds
     */
    fun remainingDuration(): Long {
        if (_currentIndex.value < 0) return 0L
        return _queue.value.drop(_currentIndex.value).sumOf { it.article.estimatedDuration }
    }

    /**
     * Check if queue contains an article
     */
    fun contains(articleId: String): Boolean {
        return _queue.value.any { it.article.id == articleId }
    }

    /**
     * Get queue item for an article
     */
    fun getQueueItem(articleId: String): QueueItem? {
        return _queue.value.find { it.article.id == articleId }
    }
}
